contributors/design-proposals/node/pod-pid-namespace.md

```diff
@@ -1,77 +1,152 @@
+*   Status: Pending
+*   Version: Alpha
+*   Implementation Owner: verb
 
 ## Motivation
 
-Sharing a PID namespace between containers in a pod is discussed in
-[#1615](https://issues.k8s.io/1615), and enables:
+Pods share namespaces where possible, but a requirement for sharing the PID
+namespace has not been defined due to lack of support in Docker. This created an
+implicit API on which certain container images now rely. This proposal
+explicitly defines behavior for sharing a PID namespace.
 
-  1. signaling between containers, which is useful for side cars (e.g. for
-     signaling a daemon process after rotating logs).
-  2. easier troubleshooting of pods.
-  3. addressing [Docker's zombie problem][1] by reaping orphaned zombies in the
-     infra container.
+## Proposal
 
-## Goals and Non-Goals
+### Goals and Non-Goals
 
 Goals include:
-  - Changing default behavior in the Docker runtime as implemented by the CRI
-  - Making Docker behavior compatible with the other Kubernetes runtimes
-
-Non-goals include:
-  - Creating an init solution that works for all runtimes
-  - Supporting isolated PID namespace indefinitely
-
-## Modification to the Docker Runtime
-
-We will modify the Docker implementation of the CRI to use a shared PID
-namespace when running with a version of Docker >= 1.12. The legacy
-`dockertools` implementation will not be changed.
 
-Linking this change to the CRI means that Kubernetes users who care to test such
-changes can test the combined changes at once. Users who do not care to test
-such changes will be insulated by Kubernetes not recommending Docker >= 1.12
-until after switching to the CRI.
+*   Backwards compatibility with container images expecting `pid == 1` semantics
+*   Per-pod configuration of PID namespace sharing
+*   Ability to change default sharing behavior in `v2.Pod`
 
-Other changes that must be made to support this change:
+Non-goals include:
 
-1. Add a test to verify all containers restart if the infra container
-   responsible for the PodSandbox dies. (Note: With Docker 1.12 if the source
-   of the PID namespace dies all containers sharing that namespace are killed
-   as well.)
-2. Modify the Infra container used by the Docker runtime to reap orphaned
-   zombies ([#36853](https://pr.k8s.io/36853)).
+*   Creating a general purpose init solution
+*   Multiple shared PID namespaces per pod
+*   Per-container configuration of PID namespace sharing
 
-## Rollout Plan
+### Configurable PID Namespace Sharing
 
-SIG Node is planning to switch to the CRI as a default in 1.6, at which point
-users with Docker >= 1.12 will receive a shared PID namespace by default.
-Cluster administrators will be able to disable this behavior by providing a flag
-to the kubelet which will cause the dockershim to revert to previous behavior.
+We will add first class support for configuring isolated and shared PID
+namespaces for pods by adding a new boolean field `sharedPID` to the pod spec,
```

Comment:
Thinking forward to a v2 API, we probably want the default to be shared, and private PID to be the exceptional case, right?  Does it make sense then to define this as `privatePID` with a default of `true` (but not `omitempty`).  Or can we imagine a plausible 3rd state (e.g. should this have been `pid: {private | shared | host}` more like Docker's native API?

This does touch on an interesting subject - defaulting in clients.  We know it's not super helpful for a client to apply defaults before sending up to the API server (since we lose track of what the user wanted vs what was defaulted).  

At some point, we opted NOT to demand Go-style "zero-value" default semantics, which means a bool with default true is legit, and a string with default non-empty is legit.  So how does a client know what the value of a field actually is?  Since optional fields are (almost) all pointers, clients that want to look at those fields can either nil-check and assume default meanings (hopefully as per our docs), or else run something we provide that applies defaults to their structs so they can use them.

But if the client understands fields that the server does not, it might default them to values and assume that the server implements them, when it doesn't.  We have some `omitempty` fields in our API - considering this (never thought about it this way) the client can't tell if a field is not implemented or just empty.

Should apiserver always be passing back fully-expanded values, such that if the client sees a nil it knows that field is not implemented on server?

I think we're sort of lost in the middle right now.

@kubernetes/sig-api-machinery-misc @lavalamp @caesarxuchao for more current thoughts on that topic.